File: LKO-ADC/koopman_project/src/normalize_data.m
function [data_norm, params] = normalize_data(data, params)
% NORMALIZE_DATA 使用Z-score方法对数据进行标准化
%   [data_norm, params] = normalize_data(data)
%   在训练模式下调用，计算并返回标准化后的数据和统计参数（均值、标准差）
%
%   data_norm = normalize_data(data, params)
%   在测试模式下调用，使用已有的统计参数来标准化新数据

if nargin < 2  % 训练模式: 计算并存储参数
    % 1. 计算每一行（每一个特征）的均值和标准差
    mu = mean(data, 2);
    sigma = std(data, 0, 2); % 沿第2维（行）计算标准差，使用n-1无偏估计

    % 2. 找到标准差为零的行（即该行所有元素都相同）
    %    为避免除以零，需要对这些行进行特殊处理
    zero_std_rows = (sigma == 0);

    % 初始化标准化数据矩阵
    data_norm = zeros(size(data));

    % 3. 对标准差不为零的"正常"行进行Z-score标准化
    valid_rows = ~zero_std_rows;
    if any(valid_rows)
        data_norm(valid_rows,:) = (data(valid_rows,:) - mu(valid_rows)) ./ sigma(valid_rows);
    end

    % 4. 标准差为零的行，标准化后为全零
    %    因为 (x - mu) 对于这些行本身就是0
    if any(zero_std_rows)
        data_norm(zero_std_rows,:) = 0;
    end
    
    % 5. 存储参数以备后续使用
    params.mu = mu;
    params.sigma = sigma;
    params.zero_std_rows = zero_std_rows;

else          % 测试模式: 使用已有的参数
    % 初始化标准化数据矩阵
    data_norm = zeros(size(data));
    
    % 1. 获取需要正常处理的行
    valid_rows = ~params.zero_std_rows;
    
    % 2. 使用训练时得到的均值和标准差进行标准化
    if any(valid_rows)
        data_norm(valid_rows,:) = (data(valid_rows,:) - params.mu(valid_rows)) ./ params.sigma(valid_rows);
    end

    % 3. 标准差为零的行同样处理为全零
    if any(params.zero_std_rows)
        data_norm(params.zero_std_rows,:) = 0;
    end
end
end

File: LKO-ADC/python_koopman_project/src/normalize_data.py
import numpy as np
import torch


def normalize_data(data, params=None):
    """
    使用均值和标准差 (z-score) 对数据进行归一化。
    如果未提供 params，则从数据中计算它们。
    (此函数保持不变)
    """
    if params is None:
        mu = np.mean(data, axis=1, keepdims=True)
        sigma = np.std(data, axis=1, keepdims=True)
        # 避免对常量特征进行零除
        sigma[sigma == 0] = 1
        params = {'mu': mu, 'sigma': sigma}

    normalized_data = (data - params['mu']) / params['sigma']
    return normalized_data, params


def denormalize_data(data, params):
    """
    使用提供的均值和sigma逆转归一化。
    此函数现在可以处理 NumPy 数组和 PyTorch Tensor。
    """
    # 首先检查输入数据的类型
    if isinstance(data, torch.Tensor):
        # 如果是 PyTorch Tensor:
        # 1. 将 numpy 格式的均值和标准差转换为 Tensor
        mu = torch.from_numpy(params['mu'])
        sigma = torch.from_numpy(params['sigma'])

        # 2. 确保 mu 和 sigma 与输入 data 在同一个设备上 (例如 a. 'cpu' 或 b. 'cuda:0')
        mu = mu.to(data.device)
        sigma = sigma.to(data.device)

        # 3. 确保数据类型一致 (例如 torch.float32)
        mu = mu.to(data.dtype)
        sigma = sigma.to(data.dtype)

        # 4. 执行反归一化操作
        return data * sigma + mu

    elif isinstance(data, np.ndarray):
        # 如果是 NumPy Array，执行原始逻辑
        return data * params['sigma'] + params['mu']

    else:
        # 处理未知类型
        raise TypeError(f"不支持的数据类型: {type(data)}. 函数只接受 numpy.ndarray 或 torch.Tensor。")